org/mozilla/jss/ssl/javax/JSSSocketChannel.java

```diff
@@ -242,53 +240,54 @@ public synchronized long read(ByteBuffer[] dsts, int offset, int length) throws
             return -1;
         }
 
-        readBuffer.clear();
-
-        long remoteRead = 0;
         long unwrapped = 0;
         long decrypted = 0;
 
         try {
+            SSLEngineResult result;
             do {
-                int buffer_size = boundRead(readBuffer.remaining());
-                ByteBuffer src = ByteBuffer.wrap(readBuffer.array(), readBuffer.position(), buffer_size);
-
-                int this_read = 0;
-                if (consumed != null) {
-                    this_read += consumedChannel.read(src);
-                } else {
-                    this_read += readChannel.read(src);
-                }
-
-                readBuffer.position(readBuffer.position() + this_read);
-                remoteRead += this_read;
-
-                if (remoteRead == 0) {
-                    return 0;
+                int n = remoteRead();
+
+                if (readBuffer.position() == 0) {
+                    // We didn't read anything and there is no left-over data.
+                    // If handshake already completed, we can continue because
+                    // there may be data in the unwrapped buffer FD that has yet
+                    // to be delivered to the application.
+                    //
+                    // But if we did not finish the handshake and there were no
+                    // new data, we have to return here.  Calling unwrap() with
+                    // no data, prior to handshake completion, breaks something.
+                    //
+                    // Note we cannot call engine.getHandshakeStatus() to perform
+                    // this check; it is NOT a passive routine and reading it
+                    // early actually causes the same breakage.
+                    //
+                    if (!handshakeCompleted) {
+                        return decrypted > 0 ? decrypted : n;
+                    }
                 }
 
                 readBuffer.flip();
 
-                SSLEngineResult result = engine.unwrap(readBuffer, dsts, offset, length);
-                if (result.getStatus() != SSLEngineResult.Status.OK &&
-                    result.getStatus() != SSLEngineResult.Status.BUFFER_UNDERFLOW &&
-                    result.getStatus() != SSLEngineResult.Status.CLOSED) {
-                    throw new IOException("Unexpected status from unwrap: " + result);
+                result = engine.unwrap(readBuffer, dsts, offset, length);
+                switch (result.getStatus()) {
+                    case CLOSED:
+                        shutdownInput();
```

Comment:
I think the same applies for write, no? We should shutdown output if the engine reports closed?

Reply:
@cipherboy probably?  Would you be willing to investigate and, if necessary, deal with that in a separate patch?

Reply:
Sure, I can do that.